### PR TITLE
Remove debug-only CPU device index assert

### PR DESCRIPTION
**Pull request by thisisatharva-rh**
### Description 
Removes the `TORCH_INTERNAL_ASSERT_DEBUG_ONLY` in `c10::Device::validate` that asserts CPU device must be -1 or 0. This assert is compiled out in release builds, so `torch.device("cpu:N")` always suceeds. 

This was exposed by #191509 which adds a debug-build CPU-only distributed CI job. About 30+ distributed test files exhibit the same pattern 

### Test Plan

```
DEBUG=1 python -m pip install -e . -v

#This no longer asserts 
python -c "import torch; print(torch.device('cpu:2'))"

#Failing tests in 191509 now pass 
python test/distributed/_shard/sharding_spec/test_sharding_spec.py TestShardingSpec.test_infer_sharding_spec_from_shards_metadata
python -m pytest test/distributed/test_device_mesh.py::InitDeviceMeshTest::test_backend_override_argument_dict_with_idx_and_backend_eager
```

cc: @d4l3k @atalman

### CODE REVIEW

**Comment by d4l3k** [comment]
@albanD @ezyang thoughts on removing this check? It's only enforced in debug mode and causing failures in distributed we can't test reasonably. In practice it doesn't matter since CPU devices get normalized when creating tensors

**Comment by albanD** [comment]
Well, I have to admit I have no idea what happens if you try to use `cpu:1` 
Does that even work? I would expect anything you try to do with that device to crash and burn.
My expectation here is that on gpu, we do these index checks consistently do we're ok with letting you create a bad device as every piece of code consistently handle the index. But I assume the cpu side is not as nicely behaved and is going to do weird thing with it?

Why are we not making this a TORCH_CHECK that will fail in all builds?

**Comment by thisisatharva-rh** [comment]
Currently we are allowing `torch.device(cpu:2)` and stuff,  so yes it works. and yes as @d4l3k said, it does seem like we are normalizing when tensors are created on such device constructions. a bunch of distributed tests pass in regular builds because of this, I currently have claude fixing those and seeing if they pass on debug builds. 

if we were to make this a TORCH_CHECK, i am not sure i have the full context of what apart from distributed tests could require changes. let me investigate

**Comment by albanD** [comment]
Do you have some context on why distributed ever creates cpu device on index non-0 ? Why is it so common?

**Comment by thisisatharva-rh** [comment]
a lot of `torch.device(device_type, self.rank)` without having a cpu guard in place. submitting a PR w those tests fixed to have an assert soon (guarding with `self.rank if device_type != "cpu" else 0"`). (30 was the number claude told me initially, but on looking further it is not as many that actually run into this issue) 
 
should we force this assert on all builds?

**Comment by thisisatharva-rh** [comment]
okay i am more convinced that we should TORCH_CHECK, https://github.com/pytorch/pytorch/pull/194612 broke nothing but the distributed tests

**Comment by d4l3k** [comment]
I believe it was disabled for performance reasons. I wonder if we can alleviate that by using branch prediction hints or only check it via explicit constructor not copy

We should fix the distributed tests regardless

**Comment by thisisatharva-rh** [comment]
Made the required test changes w the CI job since they were minimal - https://github.com/pytorch/pytorch/pull/191509, will investigate perf reasons.
